### PR TITLE
Switch parachain interface to new `runtime_interface` macro

### PR DESCRIPTION
**Pull request by bkchr**

### CODE REVIEW

**Comment by gavofyork** [comment]
It's a fairly substantial PR, so a rough overview of the refactoring would facilitate a swift review.

File: parachain/src/wasm_executor/validation_host.rs [hunk 1]
```diff
@@ -293,23 +326,23 @@ impl ValidationHost {
...
-						};
-						externalities.post_upward_message(message_ref)?;
-					}
+				ValidationResultHeader::Ok(result) => {
+					let egress = Vec::<TargetedMessage>::decode(&mut message_data)
+						.map_err(|e|
+							Error::External(
+								format!("Could not decode egress messages: {}", e.what())
+							)
+						)?;
```
Comment (thiolliere) [inline parachain/src/wasm_executor/validation_host.rs]: I struggle to understand how this egress gets written, is it validation_code responsibility to return those two vectors ? or is it written by the worker somehow at the line 158 ?
```rust
message_data[0..e_len].copy_from_slice(egress_data);
```

 I see WorkerExternalities store them in Vec now

File: test-parachains/adder/build.rs [hunk 2]
```diff
@@ -20,6 +20,6 @@ fn main() {
 	build_current_project_with_rustflags(
 		"wasm_binary.rs",
 		WasmBuilderSource::Crates("1.0.7"),
-		"-C link-arg=--import-memory",
+		"-C link-arg=--export=__heap_base",
```
Comment (thiolliere) [inline test-parachains/adder/build.rs]: I'm very bad at understanding those flags what has this changed ?

Comment (bkchr) [reply test-parachains/adder/build.rs]: Good that you found this. I wanted to open an issue for this ;) Before we provided the memory, now the WASM needs to provide it (currently default for our runtimes). However, we need to make sure that we at least provide 1GB of memory to the validation.

File: parachain/src/wasm_executor/validation_host.rs (hunk 1 above)
Comment (bkchr) [reply parachain/src/wasm_executor/validation_host.rs]: First it gets written here to a `Vec<u8>` in the validation host: https://github.com/paritytech/polkadot/pull/665/files#diff-6f0269e206acb5ff83e3c41e68c2d451R55

After the validation is finished, we write the data here into the shared memory: https://github.com/paritytech/polkadot/pull/665/files#diff-6f0269e206acb5ff83e3c41e68c2d451R158

And this code here, is the "other" side of the shared memory in the polkadot node that reads the result.

File: parachain/src/wasm_api.rs [hunk 3]
```diff
@@ -16,57 +16,50 @@
...
 
-use codec::{Encode, Decode};
-use super::{ValidationParams, ValidationResult, MessageRef, UpwardMessageRef};
+use crate::{TargetedMessage, UpwardMessage};
+use sp_runtime_interface::runtime_interface;
+#[cfg(feature = "std")]
+use sp_externalities::ExternalitiesExt;
+
+/// The parachain api for posting messages.
+#[runtime_interface]
```
Comment (rphmeier) [inline parachain/src/wasm_api.rs]: tidy: move this downwards?

File: parachain/src/wasm_api.rs [hunk 4]
```diff
@@ -16,57 +16,50 @@
...
+pub trait Parachain {
+	/// Post a message to another parachain.
+	fn post_message(&mut self, msg: TargetedMessage) {
+		self.extension::<crate::wasm_executor::ParachainExt>()
+			.expect("No `ParachainExt` associated with the current context.")
+			.post_message(msg)
+			.expect("Failed to post message")
+	}
 
-mod ll {
```
Comment (rphmeier) [inline parachain/src/wasm_api.rs]: How does the new API look from a C standpoint? Previously, it was actually easy to write a parachain in C or C++. The new API is opaque and confusing - what is `self.extension` doing on the `Parachain` trait?

File: parachain/src/wasm_executor/validation_host.rs [hunk 5]
```diff
@@ -39,6 +38,37 @@ const NUM_HOSTS: usize = 8;
...
 
+#[derive(Default)]
+struct WorkerExternalitiesInner {
+	egress_data: Vec<u8>,
+	up_data: Vec<u8>,
+}
+
+#[derive(Default, Clone)]
+struct WorkerExternalities {
+	inner: Arc<Mutex<WorkerExternalitiesInner>>,
```
Comment (rphmeier) [inline parachain/src/wasm_executor/validation_host.rs]: why does this need to be shared? I would have thought that every parachain execution can own its data completely.

File: parachain/src/wasm_executor/mod.rs [hunk 6]
```diff
@@ -84,372 +82,169 @@ pub enum Error {
...
+		#[cfg(target_os = "unknown")]
+		ExecutionMode::RemoteTest =>
+			Err(Error::System("Remote validator not available".to_string().into())),
 	}
 }
 
-impl wasmi::HostError for ExternalitiesError {}
-impl std::error::Error for ExternalitiesError {}
+/// The host functions provided by the wasm executor to the parachain wasm blob.
+type HostFunctions = (sp_io::SubstrateHostFunctions, crate::wasm_api::parachain::HostFunctions);
```
Comment (rphmeier) [inline parachain/src/wasm_executor/mod.rs]: Is this exposing all of Substrate's host functions to the parachain WASM blob? incl. storage etc?

File: parachain/src/wasm_executor/mod.rs [hunk 7]
```diff
@@ -84,372 +82,169 @@ pub enum Error {
...
+) -> Result<ValidationResult, Error> {
+	let mut ext = ValidationExternalities(ParachainExt::new(externalities));
+
+	let res = sc_executor::call_in_wasm::<_, HostFunctions>(
+		"validate_block",
+		encoded_call_data,
+		sc_executor::WasmExecutionMethod::Interpreted,
+		&mut ext,
+		validation_code,
+		// TODO: Make sure we don't use more than 1GB
```
Comment (rphmeier) [inline parachain/src/wasm_executor/mod.rs]: stray TODO needs issue filed

File: parachain/src/wasm_executor/mod.rs (hunk 6 above)
Comment (rphmeier) [reply parachain/src/wasm_executor/mod.rs]: I see a little lower down that these are all implemented but as panicking stubs. It's really weird to make the parachain WASM link to all of these and then have it never call them.

File: parachain/src/wasm_api.rs (hunk 4 above)
Comment (bkchr) [reply parachain/src/wasm_api.rs]: That is all documented in Substrate: https://crates.parity.io/sp_runtime_interface/index.html

File: parachain/src/wasm_executor/mod.rs (hunk 6 above)
Comment (bkchr) [reply parachain/src/wasm_executor/mod.rs]: Yes it is exposing all of them. 

Most of the stuff is related to the current design of Substrate and its externalities. 

I would really like to not expose the storage functions from the host, but that also requires changes to the wasm executor. The executor currently tries to resolve all extern functions and otherwise returns an error. In my opinion it should ignore functions that it does not know and error when the blob tries to execute them.

What kind of functions to we want to expose to the parachain wasm blob? I think I asked this before and the answer was that we expose the same functionality as Substrate does. So crypto functions etc. If we have support for ignoring extern functions, we can make this more fine grained, but I don't want to block on this for now.

File: parachain/src/wasm_executor/validation_host.rs (hunk 5 above)
Comment (bkchr) [reply parachain/src/wasm_executor/validation_host.rs]: They are unique, that is correct. However, the design of the externalities extensions require that we pass an owned struct that implements `Send`. (This is related to how we store them internally).

The validation context needs to access the data written by the parachain afterwards, so we wrap the inner in a `Arc<Mutex<>>` to make it clonable and to make it implement `Send`.

File: parachain/src/wasm_executor/validation_host.rs (hunk 1 above)
Comment (thiolliere) [reply parachain/src/wasm_executor/validation_host.rs]: ok got it.

but in the shared memory egress are written as an array not a vec: 
https://github.com/paritytech/polkadot/pull/665/files#diff-6f0269e206acb5ff83e3c41e68c2d451R158
```rust
message_data[0..e_len].copy_from_slice(egress_data);
```
And here it is decoded as a vec. it seems to miss the size of the vec for decoding.

Comment (bkchr) [reply parachain/src/wasm_executor/validation_host.rs]: We copy the whole encoded buffer into the shared memory (The encoded buffer is the correct SCALE encoded vec of egress messages). So, we don't miss to write the length, it's already in the data.

Comment (thiolliere) [reply parachain/src/wasm_executor/validation_host.rs]: oh yes indeed

File: parachain/src/wasm_executor/mod.rs [hunk 8]
```diff
@@ -84,372 +82,177 @@ pub enum Error {
...
-					descriptor.maximum().map(|x| memory_units::Pages(x as usize)),
-				)?;
-				*self.memory.borrow_mut() = Some(mem.clone());
-				Ok(mem)
-			}
-		} else {
-			Err(WasmError::Instantiation("Memory imported under unknown name".to_owned()))
-		}
+	fn storage_hash(&self, _: &[u8]) -> Option<Vec<u8>> {
+		panic!("Unsupported feature for parachain validation")
```
Comment (NikVolf) [inline parachain/src/wasm_executor/mod.rs]:
```suggestion
		panic!("storage_hash: unsupported feature for parachain validation")
```

File: parachain/src/wasm_executor/mod.rs [hunk 9]
```diff
@@ -84,372 +82,177 @@ pub enum Error {
...
 	}
-}
 
-struct ValidationExternals<'a, E: 'a> {
-	externalities: &'a mut E,
-	memory: &'a MemoryRef,
-}
+	fn child_storage_hash(&self, _: ChildStorageKey, _: ChildInfo, _: &[u8]) -> Option<Vec<u8>> {
+		panic!("Unsupported feature for parachain validation")
```
Comment (NikVolf) [inline parachain/src/wasm_executor/mod.rs]:
```suggestion
		panic!("child_storage_hash: unsupported feature for parachain validation")
```

**Comment by gavofyork** [comment]
Anything more to do here before we can get it merged?

**Comment by bkchr** [comment]
From my side it is ready.

File: parachain/src/lib.rs [hunk 10]
```diff
@@ -22,40 +22,40 @@
...
-//! instance and exports a function `validate`.
+//! instance and exports a function `validate_block`.
 //!
 //! `validate` accepts as input two `i32` values, representing a pointer/length pair
-//! respectively, that encodes `ValidationParams`.
+//! respectively, that encodes [`ValidationParams`].
 //!
-//! `validate` returns an `i32` which is a pointer to a little-endian 32-bit integer denoting a length.
-//! Subtracting the length from the initial pointer will give a new pointer to the actual return data,
+//! `validate` returns an `u64` which is a pointer to an `u8` array and it's length.
```
Comment (rphmeier) [inline parachain/src/lib.rs]:
```suggestion
//! `validate` returns an `u64` which is a pointer to an `u8` array and its length.
```

File: parachain/src/wasm_executor/validation_host.rs (hunk 5 above)
Comment (rphmeier) [reply parachain/src/wasm_executor/validation_host.rs]: I really hope this is on the way out soon! It's polluting a bunch of stuff.

**Comment by rphmeier** [comment]
Still not happy about the `Arc<Mutex<_>>`. It's made the externalities code really ugly and confusing. The interface doesn't make sense any more (why does it take `&mut`?) How can we get rid of that?

**Comment by bkchr** [comment]
I'm really not sure how this is more cleaner:
```
 impl ModuleImportResolver for Resolver {
	fn resolve_func(
		&self,
		field_name: &str,
		signature: &wasmi::Signature
	) -> Result<wasmi::FuncRef, WasmError> {
		match field_name {
			"ext_post_message" => {
				let index = ids::POST_MESSAGE;
				let (params, ret_ty): (&[ValueType], Option<ValueType>) =
					(&[ValueType::I32, ValueType::I32, ValueType::I32], None);

				if signature.params() != params || signature.return_type() != ret_ty {
					Err(WasmError::Instantiation(
						format!("Export {} has a bad signature", field_name)
					))
				} else {
					Ok(wasmi::FuncInstance::alloc_host(
						wasmi::Signature::new(&params[..], ret_ty),
						index,
					))
				}
			}
			"ext_upwards_post_message" => {
				let index = ids::POST_UPWARD_MESSAGE;
				let (params, ret_ty): (&[ValueType], Option<ValueType>) =
					(&[ValueType::I32, ValueType::I32], None);

				if signature.params() != params || signature.return_type() != ret_ty {
					Err(WasmError::Instantiation(
						format!("Export {} has a bad signature", field_name)
					))
				} else {
					Ok(wasmi::FuncInstance::alloc_host(
						wasmi::Signature::new(&params[..], ret_ty),
						index,
					))
				}
			}
			_ => {
				Err(WasmError::Instantiation(
					format!("Export {} not found", field_name),
				))
			}
		}

	}
...
}

impl<'a, E: 'a + Externalities> Externals for ValidationExternals<'a, E> {
	fn invoke_index(
		&mut self,
		index: usize,
		args: ::wasmi::RuntimeArgs,
	) -> Result<Option<RuntimeValue>, Trap> {
		match index {
			ids::POST_MESSAGE => self.ext_post_message(args).map(|_| None),
			ids::POST_UPWARD_MESSAGE => self.ext_post_upward_message(args).map(|_| None),
			_ => panic!("no externality at given index"),
		}
	}
}
```

as

```
#[runtime_interface]
pub trait Parachain {
	/// Post a message to another parachain.
	fn post_message(&mut self, msg: TargetedMessage) {
		self.extension::<crate::wasm_executor::ParachainExt>()
			.expect("No `ParachainExt` associated with the current context.")
			.post_message(msg)
			.expect("Failed to post message")
	}

	/// Post a message to this parachain's relay chain.
	fn post_upward_message(&mut self, msg: UpwardMessage) {
		self.extension::<crate::wasm_executor::ParachainExt>()
			.expect("No `ParachainExt` associated with the current context.")
			.post_upward_message(msg)
			.expect("Failed to post upward message")
	}
}
```

Regarding `&self mut`, I said it before and I say it again, please read the documentation: https://crates.parity.io/sp_runtime_interface/attr.runtime_interface.html

TLDR: `&self mut` will give you mutable access to the `Externalities`, `&self` shared access. This is not breaking any rust ownership conventions. It is exactly this way to make it clear what you get. 

And no, we can not get rid of it. You could propose some better interface, if you want. But it is not like that I don't have invested more than one month of work into this. To get the best possible/easy to use interface. You are right that it is a macro that hides some stuff from you, but IMHO it's better this way than maintaining 1000 lines of redundant code. If you want to see how the data is passed between the wasm side and the native side, you can look it up here: https://github.com/paritytech/substrate/blob/master/primitives/runtime-interface/src/impls.rs

For the `Arc<Mutex<>>` case, I'm with you. This is not nice. I will try to come up with something that is better.
